src/shady-css/lexer.js

```diff
@@ -0,0 +1,141 @@
+import { matcher, tokenType } from './common';
+
+class Lexer {
+  constructor(cssText) {
+    this.cssText = cssText;
+    this.offset = 0;
+  }
+
+  nextToken() {
+    let token = this.tokenAt(this.offset);
+    this.offset = token ? token.offset : this.cssText.length;
+    return token;
+  }
+
+  tokenAt(offset) {
+    let token = this.tokenizeWhitespace(offset);
```

Comment:
This is a little hard to follow here. I would expect that a call to tokenizeWhitespace means that we already know the next token is whitespace. Checking the token type right after seems confusing. Also, returning a token that's thrown away is just generating garbage to collect.

Generally the type of next token should be determinable solely from the current character (modulo some distinctions like keyword vs identifiers, which some tokenizers will distinguish and others leave to the parser). So the nextToken function would be a branch on the current char's class (testing w/ regexes), then a call to a tokenize function that always returns a token of the same type.

something like:

```
nextToken() {
  if (this.offset >= this.cssText.length) return null;
  let character = this.cssText[this.offset];
  if (matcher.whitespace.test(character)) {
    return this.tokenizeWhitespace();
  }
  if (matcher.stringBoundary.test(character) {
    return this.tokenizeString();
  }
  if (matcher.boundary.test(character) {
    return this.tokenizeBoundary();
  }
  // etc., for each token type
  // if no match, unknown character class, error
}
```